README.md

```diff
@@ -1 +1,253 @@
+      val Role = FactType[String]("Role")
+    }
+    
+    sealed trait Role
+    object Role {
+      case object Admin extends Role
+      case object User extends Role
+      
+      implicit val eq: Eq[Role] = Eq.fromUniversalEquals
+      implicit val ordering: Ordering[Role] = Ordering.by {
+        case Admin => 1
+        case User => 2
+      }.reverse
+    }
+    ```
+
+3.  **Craft your expressions.**
+
+    You must start with either a `TerminalFactsExp` or a logical operator (such as `and` / `or`). You will typically build a
+    `TerminalFactsExp` by filtering to a specific type of fact and applying a filter condition on the instances. Since every
+    `FactType` can have multiple instances, you have to decide how to filter the list with a logical rule:
+    
+    ```
+    import cats.implicits._
+    import com.rallyhealth.vapors.factfilter.dsl._
+    
+    val isAdmin = {
+      withType(FactTypes.Role).whereAnyFactValue(_ === "admin")
+    }
+    val isOver18 = {
+      withType(FactTypes.DateOfBirth).whereAnyFactValue(_ <= LocalDate.now().minusYears(18))
+    }
+    ```
+    
+4.  **Feed your facts and expression into the evaluator.** 
+
+    Assuming you have these facts:
+    
+    ```
+    import FactTypes._
+    
+    val facts = Facts(
+      DateOfBirth(LocalDate.of(1980, 1, 1)),
+      Role("user"),
+      Role("admin")
+    )
+    ```
+    
+    You can then evaluate the query to retrieve the facts that were used to prove that the query produces a sound result.  
+    
+    ```
+    import com.rallyhealth.vapors.factfilter.dsl._
+    
+    val isAdminResult = evalWithFacts(facts) {
+      isAdmin
+    }
+    assert(results.matchingFacts == List(Role("admin"))
+    
+    val combinedResult = evalWithFacts(facts) {
+      isAdmin && isOver18
+    }
+    assert(results.matchingFacts == List(DateOfBirth(LocalDate.of(1980, 1, 1)), Role("admin"))
+    ```
+    
+### Complex FactTypes
+
+You can use any type you want in the `FactType`, but in order to use comparison operators like `>`, `>=`, `<`, `<=`, and `===`
+you must define an implicit `Ordering` instance for that type.
+
+```scala
+case class FeetAndInches(feet: Int, inches: Int)
+object FeetAndInches {
+  implicit val ordering: Ordering[FeetAndInches] = Ordering.fromLessThan { (l, r) =>
+    l.feet < r.feet || (l.feet == r.feet && l.inches < r.inches)
+  }
+}
+```
+
+### Using a FactType Enumeration 
+
+You may want to put all of your `FactType` instances into an `Enumeration` for documentation and serialization purposes.
+
+**TODO: Define a better interface to extend for enumerations** 
+
+</details>
+
+## Terminology and Problem Statement
+
+<details>
+
+In essence, every query can be optimized and evaluated to produce a function: 
+```
+Facts => ResultSet
+```
+
+### Types of Expressions
+
+| Type | Description | Definition |
+| ---- | ----------- | ---------- |
+| Exp  | A generic expression. | `FreeApplicative[ExpAlg[T, *, ...], A]`* |
```

Comment:
`FreeApplicative[ExpAlg[T, *, ...], A]`* displays weird and needs more explanation probably.